crates/circuit/src/parameter_table.rs

```diff
@@ -332,26 +327,28 @@ impl ParameterTable {
                     (vector_info.refcount > 0).then_some(vector_info)
                 });
         }
-        self.order.clear();
-        self.py_parameters = None;
+        self.invalidate_cache();
         Ok(info.uses)
     }
 
     /// Clear this table, yielding the Python parameter objects and their uses in sorted order.
+    ///
+    /// The clearing effect is eager and not dependent on the iteration.
     pub fn drain_ordered(
-        &'_ mut self,
-    ) -> impl Iterator<Item = (Py<PyAny>, HashSet<ParameterUse>)> + '_ {
-        self.ensure_sorted();
+        &mut self,
+    ) -> impl ExactSizeIterator<Item = (Py<PyAny>, HashSet<ParameterUse>)> {
```

Comment:
Is there any reason to use an abstract return type here, or can we just return `ParameterTableDrain`?

In the case of `Vec::drain`, a public `Drain` struct gets returned explicitly. That seems like a useful pattern for the general case since we can then implement additional traits for the iterator if needed.

Reply:
Mostly I'd done it because I didn't want `ParameterTableDrain` to exist at all, so I kept the type private. The original version of this function had the iterator type unnameable, because I could borrow everything out of the struct without needing a manual implementation.  The Rust stdlib exports all its iterator type objects possibly in part because anonymous `impl Trait` in return position only turned up in 1.26, so there was no alternative before that.

I can make it a public type if you think it's better that way, but I'm not sure I agree with the justification "can implementation additional traits [...] if needed" in this case - if we need them, _then_ we can name the type and make it public, but at the moment, we've no need.

Reply:
Yeah, I don't know if they would have either.  The anonymous types don't make for the prettiest reading of documentation, tbf, though equally, all the structs named `Iter` just mean you have to go to a mostly empty separate docs page to find which of the `Iterator` extension traits it implements.